File: pe5.scala
import scala.math._
/*
2520 is the smallest number that can be divided by each of the numbers from 1 to 10 without any remainder.

What is the smallest positive number that is evenly divisible by all of the numbers from 1 to 20? 
*/

object pe5 {

  def bruteForceSolver = {
    var value = 100
    var evenlyDivided = false
    while (!evenlyDivided) {
      evenlyDivided = divisbleByOneThroughTwenty_?(value)
      value += 1
    }
    value = value - 1
    println("value is " + value)
  }

  def divisbleByOneThroughTwenty_?(value:Int) = {
    var divisible_? = true
    var divisor = 2
    while (divisible_? && (divisor < 21)) {
      val divided = value / divisor
      val remainder = value - divided*divisor
      if (remainder == 0)
        divisible_? = divisible_? && true
      else
        divisible_? = divisible_? && false
      divisor += 1
    }
    divisible_?
  }
}

File: pe5/pe5.go
package main

import (
	"fmt"
	"math"
)

/*
2520 is the smallest number that can be divided by each of the numbers from 1 to 10 without any remainder.

What is the smallest positive number that is evenly divisible by all of the numbers from 1 to 20?
*/

func sieve(limit int) []int {
	primes := []int{}
	list := make([]bool, limit)
	list[0] = true
	list[1] = true
	for i, isprime := range list {
		if !isprime {
			primes = append(primes, i)
			for j := i * i; j < limit; j += i {
				list[j] = true
			}
		}
	}
	return primes
}

func solve(limit int) int {
	result := 1.0
	primes := sieve(limit)
	logLimit := math.Log(float64(limit))
	for _, p := range primes {
		// The exp must be found as the maximum value such that p**x is less than the limit
		// p**x = k ==> x * log(p) = log(k) ==> x = floor( log(k) / log(p))
		exp := math.Floor(logLimit / math.Log(float64(p)))
		result = result * math.Floor(math.Pow(float64(p), exp))
	}
	return int(result)
}

func main() {
	limit := 20
	fmt.Printf("The smallest positive number evenly divisible by all of the numbers from 1 to 20 is %d\n", solve(limit))
}
